### PR TITLE
Piper/byzantium support

### PR DESCRIPTION
**Pull request by pipermerriam**
### What was wrong?

Need support for byzantium

### How was it fixed?

- Added it to the mock, pyethereum21, and pyevm backends
- New API for finding what forks are supported.
- Update to latest py-evm version.
- Changed to use consistent names (based on py-evm's naming convention) for fork names.

#### Cute Animal Picture

maxresdefault

### CODE REVIEW

File: eth_tester/utils/backend_testing.py [hunk 1]
```diff
@@ -1249,21 +1250,29 @@ def test_time_traveling_backwards_not_allowed(self, eth_tester):
...
                                              eth_tester,
                                              fork_name,
                                              expected_init_block,
                                              set_to_block):
+        if fork_name not in eth_tester.get_supported_forks():
+            with pytest.raises(UnknownFork):
+                eth_tester.get_fork_block(fork_name)
+            with pytest.raises(UnknownFork):
+                eth_tester.set_fork_block(fork_name, set_to_block)
+            return
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply eth_tester/utils/backend_testing.py]: I'm not sure what to do here.  Current solution delegates to the backend to tell us if the fork is suppored and then switch on that.  Without that check, the test would need to be aware of which backends support which forks which feels odd as it duplicates that data.

I agree that the way this test is currently written it'd be easy to get false positive test successes.  Any suggestions on specifically how we might parametrize this?

File: eth_tester/main.py [hunk 2]
```diff
@@ -130,7 +136,10 @@ def __init__(self,
         self.auto_mine_transactions = auto_mine_transactions
 
         for fork_name, fork_block in fork_blocks.items():
-            self.set_fork_block(fork_name, fork_block)
+            try:
+                self.set_fork_block(fork_name, fork_block)
+            except UnknownFork:
+                pass
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply eth_tester/main.py]: 186d7edf821f52bb599f39031d6bd8c145a0ecd4

File: eth_tester/utils/backend_testing.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (carver) [reply eth_tester/utils/backend_testing.py]: Ah yeah, I missed the issue that different backends support different forks. One approach could be to say all forks in the parameters are supposed to be supported, but then have per-backend escapes, like:

```py
# pseudocode
if eth_tester.backend == pyethereum16 and fork_name in {
        FORK_TANGERINE_WHISTLE, FORK_BYZANTIUM
    }:
        pytest.mark.xfail(reason="pyethereum 1.6 does not have support for fork %s" % fork_name)
elif eth_tester.backend == pyethereum21 and fork_name in {
        FORK_BYZANTIUM
    }:
        pytest.mark.xfail(reason="pyethereum 2.1 does not have support for fork %s" % fork_name)
```

[the remark this answers is not in the document]
Comment (pipermerriam) [reply eth_tester/utils/backend_testing.py]: 1c539b2be77b9a20485d34d9125fcbc31d9d9152

**Pull request merged**